arangod/IResearch/IResearchPrimaryKeyFilter.h

```diff
@@ -0,0 +1,149 @@
+    irs::sub_reader const& segment,
+    irs::order::prepared const& /*order*/,
+    irs::attribute_view const& /*ctx*/
+  ) const override;
+
+// ----------------------------------------------------------------------------
+// --SECTION--                                                      irs::filter
+// ----------------------------------------------------------------------------
+
+  virtual size_t hash() const noexcept override;
+
+  virtual filter::prepared::ptr prepare(
+    irs::index_reader const& index,
+    irs::order::prepared const& /*ord*/,
+    irs::boost::boost_t /*boost*/,
+    irs::attribute_view const& /*ctx*/
+  ) const override;
+
+ protected:
+  bool equals(filter const& rhs) const noexcept override;
+
+ private:
+  struct PrimaryKeyIterator final : public irs::doc_iterator {
+    PrimaryKeyIterator() = default;
+
+    virtual bool next() noexcept override {
+      _doc = _next;
+      _next = irs::type_limits<irs::type_t::doc_id_t>::eof();
+      return !irs::type_limits<irs::type_t::doc_id_t>::eof(_doc);
+    }
+
+    virtual irs::doc_id_t seek(irs::doc_id_t target) noexcept override {
+      _doc = target <= _next
+        ? _next
+        : irs::type_limits<irs::type_t::doc_id_t>::eof();
+
+      return _doc;
+    }
+
+    virtual irs::doc_id_t value() const noexcept override {
+      return _doc;
+    }
+
+    virtual irs::attribute_view const& attributes() const noexcept override {
+      return irs::attribute_view::empty_instance();
+    }
+
+    void reset(irs::sub_reader const& segment, irs::doc_id_t doc) noexcept {
+      _pkSegment = &segment;
+      _doc = irs::type_limits<irs::type_t::doc_id_t>::invalid();
+      _next = doc;
+    }
+
+    mutable irs::sub_reader const* _pkSegment{};
+    mutable irs::doc_id_t _doc{ irs::type_limits<irs::type_t::doc_id_t>::invalid() };
+    mutable irs::doc_id_t _next{ irs::type_limits<irs::type_t::doc_id_t>::eof() };
+  }; // PrimaryKeyIterator
+
+  DocumentPrimaryKey _pk;
+  mutable PrimaryKeyIterator _pkIterator;
+}; // PrimaryKeyFilter
+
+///////////////////////////////////////////////////////////////////////////////
+/// @class PrimaryKeyFilterContainer
+/// @brief container for storing 'PrimaryKeyFilter's, does nothing as a filter
+///////////////////////////////////////////////////////////////////////////////
+class PrimaryKeyFilterContainer final : public irs::empty {
+ public:
+  DECLARE_FILTER_TYPE();
+
+  PrimaryKeyFilterContainer() = default;
+  PrimaryKeyFilterContainer(PrimaryKeyFilterContainer&&) = default;
+  PrimaryKeyFilterContainer& operator=(PrimaryKeyFilterContainer&&) = default;
+
+  PrimaryKeyFilter& emplace(TRI_voc_cid_t cid, TRI_voc_rid_t rid) {
+    _filters.emplace_back(cid, rid);
+    return _filters.back();
+  }
+
+  bool empty() const noexcept {
+    return _filters.empty();
+  }
+
+  void clear() noexcept {
+    _filters.clear();
+  }
+
+ private:
+  std::deque<PrimaryKeyFilter> _filters; // pointers remain valid
```

Comment:
Not sure what this container is used for exactly.
But it looks like the only methods that are called on the underlying container (deque) are `empty()`, `clear()`, `emplace_back()` and `back()`. If the references returned by `back()` are short-lived it looks much more appropriate to use an `std::vector` as the underlying container. An `std::vector` in reality has much lower setup costs than an `std::deque` in most STL implementations. However, it's all about iterator invalidation here, and I don't know how long the references returned by `emplace()` must remain valid.